hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java

```diff
@@ -3727,14 +3734,24 @@ public ScanResponse scan(final RpcController controller, final ScanRequest reque
         if (context != null) {
           context.setCallBack(rsh.shippedCallback);
         } else {
-          // When context != null, adding back the lease will be done in callback set above.
-          addScannerLeaseBack(lease);
+          // If context is null,we call rsh.shippedCallback directly to release the
+          // internal resources in rsh.
+          runShippedCallback(rsh);
```

Comment:
The issue here is scanner closed too early then the result bytebuffer might be overwritten. I think the scanner resources are released and lease is added back too, so is it necessary to change codes here?And if there is no rpc call context here, is it a little strange to run a kind of RpcCallBack?

Reply:
Oh, it also fixed the scanner leak problem caused by the null context and partial results. LGTM.